service/handlers.go

```diff
@@ -141,12 +116,20 @@ func GetStoreInfo(c *gin.Context) {
 	c.JSON(http.StatusOK, store)
 }
 
-func Cache(c *gin.Context) {
+func GetActivities(c *gin.Context) {
+	userActivityManager, _ := datastore.UserActivityFactory()
 	username := c.Query("username")
-	actions, errCache := CacheGet(username)
-	if errCache != nil {
-		c.JSON(http.StatusBadRequest, errCache)
+	actions, err := userActivityManager.Get(username)
+	if err != nil {
+		c.JSON(http.StatusBadRequest, err)
 		return
 	}
 	c.JSON(http.StatusOK, actions)
 }
+
+func middleware(c *gin.Context) {
+	userActivityManager, _ := datastore.UserActivityFactory()
+	username := c.Query("username")
```

Comment:
check if the username != ""

Reply:
I check it on the implementation of AddActivity, but I added the check anyway

Reply:
it is better not to call the function (in your case `AddActivity`) if the evalidation fails